libarchive/archive_write_set_format_mtree.c

```diff
@@ -1754,134 +1753,133 @@ mtree_entry_setup_filenames(struct archive_write *a, struct mtree_entry *file,
 #else
 	(void)a; /* UNUSED */
 #endif
-	pathname =  file->pathname.s;
+	pathname = file->pathname.s;
 	if (strcmp(pathname, ".") == 0) {
 		archive_strcpy(&file->basename, ".");
 		return (ARCHIVE_OK);
 	}
 
-	archive_strcpy(&(file->parentdir), pathname);
-
-	len = file->parentdir.length;
-	p = dirname = file->parentdir.s;
-
-	/*
-	 * Remove leading '/' and '../' elements
-	 */
-	while (*p) {
-		if (p[0] == '/') {
-			p++;
+	/* Remove leading '/' and '../' elements */
+	len = archive_strlen(&file->pathname);
```

Comment:
Move the size computation after the loop, e.g. archive_strlen(...) - (pathname -> file->pathname.s) or so.